core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -59,24 +61,101 @@ object RequestChannel extends Logging {
 
+    // map[acks, map[requestSize, metricName]]
+    // this is used to create metrics for produce requests of different size and acks
+    val produceRequestAcksSizeMetricNameMap = getProduceRequestAcksSizeMetricNameMap()
+    // map[responseSize, metricName]
+    // this is used to create metrics for fetch requests of different response size
+    val consumerFetchRequestSizeMetricNameMap = getConsumerFetchRequestAcksSizeMetricNameMap
+
     (enabledApis.map(_.name) ++
       Seq(RequestMetrics.MetadataAllTopics) ++
-      Seq(RequestMetrics.consumerFetchMetricName, RequestMetrics.followFetchMetricName)).foreach { name =>
-      metricsMap.put(name, new RequestMetrics(name))
+      Seq(RequestMetrics.consumerFetchMetricName, RequestMetrics.followFetchMetricName) ++
+      getConsumerFetchRequestSizeMetricNames ++
+      getProduceRequestAcksSizeMetricNames)
+      .foreach { name => metricsMap.put(name, new RequestMetrics(name))
     }
 
     def apply(metricName: String): RequestMetrics = metricsMap(metricName)
 
     def close(): Unit = {
        metricsMap.values.foreach(_.removeMetrics())
     }
+
+    // generate map[request/responseSize, metricName] for requests of different size
+    def getRequestSizeMetricNameMap(requestType: String):Map[Int, String] = {
+      // convert requestMetricsSizeBucketsConfig from string to int array. e.g., "0,1,10,50,100" => [0,1,10,50,100]
+      var bucketsString = requestMetricsSizeBucketsConfig.replaceAll("\\s", "").split(',')
+      if(bucketsString.length < 2) {
+        error(s"requestMetricsSizeBuckets has fewer than 2 buckets: '${requestMetricsSizeBucketsConfig}'. " +
+          s"use default value instead ")
+        bucketsString = Defaults.RequestMetricsSizeBuckets.replaceAll("\\s", "").split(',')
+      }
+      var bucketsInt = Array[Int](bucketsString.length)
+      try {
+        bucketsInt = bucketsString.map(_.toInt)
+      } catch {
+        case e: Exception => {
+          error(s"Failed to parse requestMetricsSizeBuckets: '${requestMetricsSizeBucketsConfig}'. " +
+            s"use default value instead ", e)
+          bucketsInt = Defaults.RequestMetricsSizeBuckets.replaceAll("\\s", "").split(',').map(_.toInt)
+        }
+      }
+
+      // get size and corresponding term to be used in metric name
+      // [0,1,10,50,100] => requestSizeBuckets:Seq((0, "0To1Mb"), (1, "1To10Mb"), (10, "10To50Mb"), (50, "50To100Mb"),
+      // (100, "100MbGreater"))
+      val requestSizeBuckets = for (i <- 0 until bucketsInt.length) yield {
+        val size = bucketsInt(i)
+        if (i == bucketsInt.length - 1) (size, s"${size}MbGreater")
+        else (size, s"${size}To${bucketsInt(i + 1)}Mb")
+      }
+      requestSizeBuckets.map(bucket => (bucket._1, s"${requestType}${bucket._2}")).toMap
+    }
+
+    // generate map[acks, map[requestSize, metricName]] for produce requests of different request size and acks
+    def getProduceRequestAcksSizeMetricNameMap():Map[Int, Map[Int, String]] = {
```

Comment:
It seems the 4 methods starting from this one can be made "private" to this class.

Reply:
I did not find a good way to test private methods in scala, so have these as public methods